### PR TITLE
Service manager fixes

### PR DESCRIPTION
**Pull request by smashery**
This resolves several edge-case bugs in the service manager. I've grouped them together into one issue because the fixes are in some cases overlapping, and it seems better to test the entire feature at once given all the changes to it.

Bug 1: Communication channel not respected in service manager #16314

- [ ]  Run an HTTP handler with the same settings (0.0.0.0 and the same port) locally, and on a remote session using ReverseListenerComm. Verify that listeners are started both locally, and via the remote session.
- [ ] Run an HTTP handler with the same settings (0.0.0.0 and the same port) on two separate remote sessions using ReverseListenerComm. Verify that listeners are started via both remote sessions.

Bug 2: Socket leak: HTTP service not dereferenced #16315

Verify the following:

- `use exploit/multi/handler`
- `set payload linux/x64/meterpreter_reverse_http`
- `set lport 8080`
- `set lhost 0.0.0.0`
- `run -j`
- Verify: netstat should show port 8080 open
- `set luri abc`
- `run -j`
- Verify: Should launch a new handler; should not complain about "the port is already open"

Bug 3: String key collision

The service manager attempts to find existing services based on a string-concatenated key of important properties (i.e. IP, port number). No delimiter was placed between the port and the host, so Port 123 on host 45.6.7.8 would be treated the same as port 1234 on host 5.6.7.8, and the service. I can't think of a situation that it's likely to happen (given that it actually needs to bind to that IP), but it's definitely a logic error. You can probably just verify this with a code review.

Bug 4: Modules don't use reference counting

The modules `exploit/multi/misc/java_rmi_server` and `exploit/osx/browser/safari_file_policy` call `stop` on the service directly, rather than respecting reference counting. As a result, any other services running at the same time as them on the same port will be killed by these modules.

### CODE REVIEW

File: lib/rex/post/meterpreter/packet_dispatcher.rb [hunk 1]
```diff
@@ -729,9 +732,7 @@ def shutdown_passive_dispatcher
       resource_uri = "/" + self.conn_id.to_s.gsub(/(^\/|\/$)/, '') + "/"
       self.passive_service.remove_resource(resource_uri) if self.passive_service
 
-      if self.passive_service.resources.empty?
-        Rex::ServiceManager.stop_service(self.passive_service)
-      end
+      Rex::ServiceManager.stop_service(self.passive_service)
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/packet_dispatcher.rb:735]: Just confirming, should this be a `deref` call rather than an explicit call to `stop_manager` ? 👀

Comment (adfoster-r7) [reply lib/rex/post/meterpreter/packet_dispatcher.rb:735]: Two http sessions opened:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > sessions

Active sessions
===============

  Id  Name  Type                 Information               Connection
  --  ----  ----                 -----------               ----------
  1         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()
  2         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()

```

services:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > _servicemanager 
Services
========

 Id  Name                                         References
 --  ----                                         ----------
 0   HTTP Server                                  4
 1   Rex::Proto::Http::Server8080-192.168.123.1-  4
```

Killing the first session:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > sessions -k 1
[*] Killing the following session(s): 1
[*] Killing session 1

[*]  - Meterpreter session 1 closed.
```

Confirming first session killed:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > sessions

Active sessions
===============

  Id  Name  Type                 Information               Connection
  --  ----  ----                 -----------               ----------
  2         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()
```

But now the service manager is empty:

```
msf6 payload(osx/x64/meterpreter_reverse_http) > _servicemanager 
[*] No framework services are currently running.
```

the listener is still open:
```
lsof -i :8080
COMMAND   PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
ruby    23249 a   14u  IPv4 0xea596340b33920c7      0t0  TCP 192.168.123.1:http-alt (LISTEN)
```

And it's no longer possible to create a new handler:
```
msf6 payload(osx/x64/meterpreter_reverse_http) > to_handler
[*] Payload Handler Started as Job 1
msf6 payload(osx/x64/meterpreter_reverse_http) > 
[-] Handler failed to bind to 192.168.123.1:8080
[*] Started HTTP reverse handler on http://0.0.0.0:8080/base_url
```

Comment (adfoster-r7) [reply lib/rex/post/meterpreter/packet_dispatcher.rb:735]: Interestingly even though the service manager claims to be empty with no services, you can still open new sessions:

```
msf6 payload(osx/x64/meterpreter_reverse_http) > _servicemanager 
[*] No framework services are currently running.
```

Opening a new payload continues to work:

```
msf6 payload(osx/x64/meterpreter_reverse_http) > [*] Meterpreter session 3 opened (192.168.123.1:8080 -> 127.0.0.1 ) at 2022-03-09 11:39:19 +0000
```

File: lib/msf/core/handler/reverse_http.rb [hunk 2]
```diff
@@ -260,9 +260,7 @@ def setup_handler
   def stop_handler
     if self.service
       self.service.remove_resource((luri + "/").gsub("//", "/"))
-      if self.service.resources.empty? && self.sessions == 0
-        Rex::ServiceManager.stop_service(self.service)
-      end
+      Rex::ServiceManager.stop_service(self.service)
```
Comment (adfoster-r7) [inline lib/msf/core/handler/reverse_http.rb:263]: Similar to the below question, should this be something akin to:

```suggestion
      self.service.deref if self.service
      self.service = nil
```

**Comment by sempervictus** [comment]
@adfoster-r7 - i would venture those "additional issues" come from the circle of hell where we keep the handler/session logic (they made a new one, just for us - number 1337).
Overall, i think the deref approach is how Rex intended this to work - though i'm a bit behind at this point after the recent changes on whether these are refs right now or not (apologies, some messes being made requiring attention).
Personally, i think that all service instances should be refcounted with a final deref calling #close. I will try to carve some time in the next few days to dive the issues/PRs currently up for this stuff, but given everything going on i may not be able to put in the requisite time... I suppose its better to fix some stuff now and not let perfect be the enemy of good, i'm just not sure if all of the changes going in follow the same design/architecture and about related knock-on effects.

**Comment by smashery** [comment]
Yeah, I think some of these issues are related to how modules think `stop_service` works... maybe there's some vestigial stuff in there whereby it used to actually stop the service, but that isn't the case now - instead, it just does a dereference. I'll neaten it up to be clearer what the intent of each function is.

File: lib/rex/post/meterpreter/packet_dispatcher.rb (hunk 1 above)
Comment (smashery) [reply lib/rex/post/meterpreter/packet_dispatcher.rb:735]: Right, this turned out to be a weirdness in the service manager; I'd initially kept the existing convention of services sayping "I'm done" by letting the `ServiceManager` know; but that just had the effect that the ServiceManager no longer kept track of it; but the service was indeed still running. Trying to start a new one would then fail because the ServiceManager no longer knew that it was running; although the deref would at least still clean it up correctly. I've now changed it to just use deref everywhere, and it seems to be behaving a lot more consistently.

**Comment by smashery** [comment]
Okay, so looking more into this, it became apparent that this is an issue in a few other places.

`exploit/multi/misc/java_rmi_server` unilaterally shuts down the HTTP server once it's done; even if something else was using it. Here I start an HTTP_NTLM listener, and then the `java_rmi_server`, and then shut the latter down; the former is also killed in doing so:

```
msf6 > use auxiliary/server/capture/http_ntlm 
msf6 auxiliary(server/capture/http_ntlm) > run
[*] Auxiliary module running as background job 0.
msf6 auxiliary(server/capture/http_ntlm) > 
[*] Using URL: http://0.0.0.0:8080/aXuT3BU
[*] Local IP: http://192.168.20.206:8080/aXuT3BU
[*] Server started.

msf6 auxiliary(server/capture/http_ntlm) > use exploit/multi/misc/java_rmi_server 
[*] No payload configured, defaulting to java/meterpreter/reverse_tcp
msf6 exploit(multi/misc/java_rmi_server) > jobs

Jobs
====

  Id  Name                                 Payload  Payload opts
  --  ----                                 -------  ------------
  0   Auxiliary: server/capture/http_ntlm

msf6 exploit(multi/misc/java_rmi_server) > set rhosts 127.0.0.1
rhosts => 127.0.0.1
msf6 exploit(multi/misc/java_rmi_server) > run

[*] Started reverse TCP handler on 192.168.20.206:4444 
[*] 127.0.0.1:1099 - Using URL: http://0.0.0.0:8080/6dHKzvaROOlXeOQ
[*] 127.0.0.1:1099 - Local IP: http://192.168.20.206:8080/6dHKzvaROOlXeOQ
[*] 127.0.0.1:1099 - Server started.
[-] 127.0.0.1:1099 - Exploit failed [unreachable]: RuntimeError The connection was refused by the remote host (127.0.0.1:1099).
[*] Exploit completed, but no session was created.
msf6 exploit(multi/misc/java_rmi_server) > 
[*] Server stopped.

msf6 exploit(multi/misc/java_rmi_server) > jobs

Jobs
====

No active jobs.
```

The `exploit/osx/browser/safari_file_policy` module had the same problem too.

**Comment by jmartin-r7** [comment]
@msjenkins-r7 test this please.

**Comment by sempervictus** [comment]
I recently mucked with how stops work, because some modules run multiple services, so i split them out. We might be looking at the unintended effects of that. @smashery, if you have a sec, could you possibly trace the stop calls in there and see whether its an early call from the module or something tripping in the actual lib piece? I didn't audit all consumers, so this may well be on me.

**Comment by smashery** [comment]
@sempervictus I realise I've combined a few issues into this one PR here, which is muddying the waters a bit. I think what I've seen in general is just a bit of inconsistency in how different parts of the code interact with the service manager: 

- The reverse HTTP handler didn't really use reference counting; rather, keeping track of how many sessions and resources existed... but those resources could have been from other modules, which themselves did respect the dereferencing. So it was a mix of conventions that were getting in each others' ways.
- When it did stop, the reverse HTTP handler would directly tell the service manager to stop, which would dereference it; but if other services were still using it, then the service manager would lose its own knowledge of the service, but the service wouldn't die, so the port would be open and the ServiceManager would be unable to find it for other modules wanting to use the service on that port, so the binding would fail.
- Two modules would just kill the service (Bug 4)

I think it's at a point where it's more consistent, and I can't see any issues (since fixing the one @adfoster-r7 raised).

Are you suggesting, though, that you have some outstanding work that might conflict with these changes?

**Comment by adfoster-r7** [comment]
@msjenkins-r7 test this please

File: lib/msf/core/handler/reverse_http.rb [hunk 3]
```diff
@@ -260,9 +260,7 @@ def setup_handler
   def stop_handler
     if self.service
       self.service.remove_resource((luri + "/").gsub("//", "/"))
-      if self.service.resources.empty? && self.sessions == 0
-        Rex::ServiceManager.stop_service(self.service)
-      end
+      self.service.deref
```
Comment (adfoster-r7) [inline lib/msf/core/handler/reverse_http.rb:263]:
```suggestion
      self.service.deref
      self.service = nil
```

File: modules/auxiliary/gather/safari_file_url_navigation.rb [hunk 4]
```diff
@@ -322,7 +322,7 @@ def cleanup
     super
 
     # Kill FTP
-    stop_service
+    cleanup_service
 
     # clear my resource, deregister ref, stop/close the HTTP socket
     begin
```
Comment (adfoster-r7) [inline modules/auxiliary/gather/safari_file_url_navigation.rb:328]: Looks like this has explicit calls to `@http_service.stop` and `@http_server.close`, I feel like that's something that we can remove now? 👀

Comment (adfoster-r7) [reply modules/auxiliary/gather/safari_file_url_navigation.rb:328]: Running this module as well as a http handler together on the same ip/port:
```
msf6 auxiliary(gather/safari_file_url_navigation) > jobs

Jobs
====

  Id  Name                                          Payload                           Payload opts
  --  ----                                          -------                           ------------
  6   Auxiliary: gather/safari_file_url_navigation
  7   Exploit: multi/handler                        osx/x64/meterpreter_reverse_http  http://192.168.123.1:8080/base_url
```

Getting sessions:
```
msf6 auxiliary(gather/safari_file_url_navigation) > sessions 

Active sessions
===============

  Id  Name  Type                 Information               Connection
  --  ----  ----                 -----------               ----------
  2         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()
  3         meterpreter x64/osx  a @ 192.168.1.239  192.168.123.1:8080 -> 127.0.0.1  ()


```

Killing the safari file module:
```
msf6 auxiliary(gather/safari_file_url_navigation) > jobs -k 6
[*] Stopping the following job(s): 6
[*] Stopping job 6

[*] Server stopped.
```

Verifying jobs:
```
msf6 auxiliary(gather/safari_file_url_navigation) > jobs

Jobs
====

  Id  Name                    Payload                           Payload opts
  --  ----                    -------                           ------------
  7   Exploit: multi/handler  osx/x64/meterpreter_reverse_http  http://192.168.123.1:8080/base_url
```

Interacting with a previously opened session is now dead:

```
msf6 auxiliary(gather/safari_file_url_navigation) > sessions -i -1 -C dir
[*] Running 'dir' on meterpreter session 3 ()


[-] Error running command dir: Rex::TimeoutError Operation timed out.
```

File: modules/exploits/linux/http/vestacp_exec.rb [hunk 5]
```diff
@@ -239,10 +239,13 @@ def payload_implant
...
-    login
-    start_backup_and_trigger_payload
-    stop_service
+    begin
+      payload_implant
+      login
+      start_backup_and_trigger_payload
+    ensure
+      cleanup_service
+    end
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/vestacp_exec.rb:248]: If `cleanup_service` gets called by default as part of `remote/socket_server`, I believe this isn't required?

```
>> framework.modules.create('exploits/linux/http/vestacp_exec').class.ancestors.map { |x| x.instance_method(:cleanup).source_location rescue nil }.compact
=> 
[["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/file_dropper.rb", 110],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/file_dropper.rb", 110],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/remote/http_server.rb", 228],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/remote/http_client.rb", 342],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/remote/socket_server.rb", 59],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/remote/socket_server.rb", 59],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/remote/tcp.rb", 200],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit/remote/tcp.rb", 200],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit.rb", 423],
 ["/Users/a/Documents/code/metasploit-framework/lib/msf/core/exploit.rb", 423]]
>> 

```

**Comment by adfoster-r7** [comment]
Ran through a few scenarios, and this does a great job of cleaning up things as expected now. I imagine there might be gremlins still, but after the feedback is applied I think this should be good to ship ⛵

File: modules/exploits/linux/http/vestacp_exec.rb (hunk 5 above)
Comment (smashery) [reply modules/exploits/linux/http/vestacp_exec.rb:248]: Yes, I think I'd made a bad assumption about what calls what; this shouldn't be needed.

File: modules/exploits/multi/http/ubiquiti_unifi_log4shell.rb [hunk 6]
```diff
@@ -150,6 +150,6 @@ def exploit
     wait_until { @search_received && (!handler_enabled? || session_created?) }
     handler
   ensure
-    cleanup
+    cleanup_service
```
Comment (smashery) [inline modules/exploits/multi/http/ubiquiti_unifi_log4shell.rb:153]: I'd guess this one isn't needed then either @adfoster-r7 - the exploit driver should clean it up. Although the one above in `check` may be needed... exploits' `check` runner doesn't seem to call cleanup:

https://github.com/rapid7/metasploit-framework/blob/c3465a8ad8369ef2eab3acd9d1a21b19feed4d3e/lib/msf/base/simple/exploit.rb#L212

File: lib/msf/core/exploit/remote/http_server/php_include.rb [hunk 7]
```diff
@@ -62,7 +62,7 @@ def exploit
     rescue ::Interrupt
       raise $!
     ensure
-      stop_service
+      cleanup_service
```
Comment (smashery) [inline lib/msf/core/exploit/remote/http_server/php_include.rb:65]: Can't see this needed either, given the exploit driver does this:

https://github.com/rapid7/metasploit-framework/blob/c3465a8ad8369ef2eab3acd9d1a21b19feed4d3e/lib/msf/core/exploit_driver.rb#L180-L190

File: modules/auxiliary/gather/safari_file_url_navigation.rb (hunk 4 above)
Comment (smashery) [reply modules/auxiliary/gather/safari_file_url_navigation.rb:328]: Thanks - I've removed the `stop` and `close`, and let `deref` do its thing.

**Comment by adfoster-r7** [comment]
## Release Notes

Ensures individual modules no longer accidentally shutdown joint services that are used across multiple modules/handlers etc, such as HTTP servers. Modules will now correctly unregister interest in the global service, and if there are no longer any interested modules in the running global service, it will be shut down correctly.
